### PR TITLE
Fix Expressions, Allow expressions to be selected, Add function expressions, string functions

### PR DESCRIPTION
**Pull request by MattPerron**

### CODE REVIEW

**Comment by MattPerron** [comment]
@malin1993ml I know there is a lot of code here and that this PR doesn't fix everything, but can you take a look?

**Comment by malin1993ml** [comment]
Sure. I'll take a look.

File: src/parser/sql_parser.y [hunk 1]
```diff
@@ -791,24 +807,23 @@ num_literal:
 	;
 
 int_literal:
-		INTVAL { $$ = new peloton::expression::ConstantValueExpression(peloton::common::ValueFactory::GetIntegerValue($1)); $$->ival = $1; }
+		INTVAL { $$ = new peloton::expression::ConstantValueExpression(peloton::common::ValueFactory::GetIntegerValue($1)); $$->ival_ = $1; }
```
Comment (eric-haibin-lin) [inline src/parser/sql_parser.y]: Is ival_ actually a public member variable?

File: src/include/expression/parser_expression.h [hunk 2]
```diff
@@ -1,116 +0,0 @@
...
-
-  ParserExpression(ExpressionType type, char* func_name_,
-                   AbstractExpression* expr_, bool distinct_)
-      : AbstractExpression(type) {
-    name = func_name_;
-    expr = expr_;
-    distinct = distinct_;
-  }
-
-  ParserExpression(ExpressionType type, int placeholder)
```
Comment (eric-haibin-lin) [inline src/include/expression/parser_expression.h]: I'm so happy to see this file deleted

File: src/parser/sql_parser.y (hunk 1 above)
Comment (MattPerron) [inline src/parser/sql_parser.y]: Yes.

File: src/include/parser/statement_prepare.h [hunk 3]
```diff
@@ -14,7 +14,7 @@
 
 #include "parser/sql_statement.h"
 #include "parser/statement_select.h"
-#include "expression/abstract_expression.h"
+#include "expression/parameter_value_expression.h"
 
 #include <algorithm>
```
Comment (malin1993ml) [inline src/include/parser/statement_prepare.h]: I actually don't the purpose of this class. Do you know when will we use it?

File: src/optimizer/simple_optimizer.cpp [hunk 4]
```diff
@@ -232,12 +279,12 @@ std::shared_ptr<planner::AbstractPlan> SimpleOptimizer::BuildPelotonPlanTree(
...
       // Else, do aggregations on top of scan
       else {
-        // Create sequential scan plan
-        target_table = catalog::Catalog::GetInstance()->GetTableWithName(
-            select_stmt->from_table->GetDatabaseName(),
-            select_stmt->from_table->GetTableName());
+//        // Create sequential scan plan
+//        target_table = catalog::Catalog::GetInstance()->GetTableWithName(
+//            select_stmt->from_table->GetDatabaseName(),
+//            select_stmt->from_table->GetTableName());
```
Comment (malin1993ml) [inline src/optimizer/simple_optimizer.cpp]: Delete unused code.

File: src/optimizer/simple_optimizer.cpp [hunk 5]
```diff
@@ -130,20 +136,60 @@ std::shared_ptr<planner::AbstractPlan> SimpleOptimizer::BuildPelotonPlanTree(
...
 
       // If there is no aggregate functions, just do a sequential scan
-      if (!func_flag && group_by_columns.size() == 0) {
+      if (!agg_flag && group_by_columns.size() == 0) {
         LOG_TRACE("No aggregate functions found.");
-        auto child_SelectPlan = CreateScanPlan(target_table, select_stmt);
+        std::unique_ptr<planner::AbstractPlan> child_SelectPlan = CreateScanPlan(target_table, column_ids, predicate, select_stmt->is_for_update);
+
+
+        if (needs_projection){
```
Comment (malin1993ml) [inline src/optimizer/simple_optimizer.cpp]: Can you add some comments to the logic here?

File: src/include/parser/statement_prepare.h (hunk 3 above)
Comment (MattPerron) [inline src/include/parser/statement_prepare.h]: Yes this is used for prepared statments.
See https://www.postgresql.org/docs/9.3/static/sql-prepare.html

Although I don't know if it really works. I have never tried it.

Comment (malin1993ml) [inline src/include/parser/statement_prepare.h]: Okay. It seems never being used in the system, either.

**Comment by MattPerron** [comment]
@malin1993ml @eric-haibin-lin 
Alright to merge?

**Comment by malin1993ml** [comment]
@MattPerron Merged.

**Comment by malin1993ml** [comment]
I tried to run TPC-C with OLTPBench after I merged this PR. If I only run NewOrder then everything works fine. But I got zero throughput when running the full TPC-C. This is the error I got:

> [ERROR] [/home/lin/peloton/src/networking/protocol.cpp:773:ProcessPacket()] 2016-11-19 00:36:51 - Packet type not supported yet: 67 (C)
> [ERROR] [/home/lin/peloton/src/networking/protocol.cpp:773:ProcessPacket()] 2016-11-19 00:36:57 - Packet type not supported yet: 67 (C)
> [ERROR] [/home/lin/peloton/src/tcop/tcop.cpp:226:GetColumnFieldForValueType()] 2016-11-19 00:37:08 - Unrecognized column type: 0
> [ERROR] [/home/lin/peloton/src/tcop/tcop.cpp:226:GetColumnFieldForValueType()] 2016-11-19 00:37:08 - Unrecognized column type: 0
> [ERROR] [/home/lin/peloton/src/tcop/tcop.cpp:226:GetColumnFieldForValueType()] 2016-11-19 00:37:08 - Unrecognized column type: 0
> [ERROR] [/home/lin/peloton/src/tcop/tcop.cpp:226:GetColumnFieldForValueType()] 2016-11-19 00:37:08 - Unrecognized column type: 0
> [ERROR] [/home/lin/peloton/src/tcop/tcop.cpp:226:GetColumnFieldForValueType()] 2016-11-19 00:37:08 - Unrecognized column type: 0

I was able to run TPC-C with the master version before this PR (930a5b3c5b6071290a106b107dbc09b242eab0b5). @MattPerron Can you have a look on this?

**Comment by MattPerron** [comment]
@malin1993ml see #351

**Pull request merged**